larpandoracontent/LArObjects/LArAdaBoostDecisionTree.cc

```diff
@@ -0,0 +1,299 @@
+                "BoostWeight", boostWeight));
+        }
+        else if ("Node" == pHeadTiXmlElement->ValueStr())
+        {
+            pandora::TiXmlHandle nodeHandle(pHeadTiXmlElement);
+            this->ReadNode(nodeHandle, idToNodeMap);
+        }
+    }
+
+    const WeakClassifier *pWeakClassifier = new WeakClassifier(idToNodeMap, boostWeight, treeIndex);
+    m_weakClassifiers.push_back(pWeakClassifier);
+
+    return pandora::STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+pandora::StatusCode AdaBoostDecisionTree::ReadNode(pandora::TiXmlHandle &currentHandle, IDToNodeMap &idToNodeMap)
+{
+    int nodeID(-1);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "NodeID", nodeID));
+
+    int variableID(-1);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "VariableID", variableID));
+
+    double threshold(0.);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "Threshold", threshold));
+
+    int leftDaughterID(-1);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "LeftDaughterID", leftDaughterID));
+
+    int rightDaughterID(-1);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "RightDaughterID", rightDaughterID));
+
+    int parentID(-1);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "ParentID", parentID));
+
+    bool outcome(false);
+    PANDORA_RETURN_RESULT_IF_AND_IF(pandora::STATUS_CODE_SUCCESS, pandora::STATUS_CODE_NOT_FOUND, !=, pandora::XmlHelper::ReadValue(currentHandle,
+        "Outcome", outcome));
+
+    bool isLeaf(leftDaughterID == -1 ? true : false);
+
+    const Node *pNode = new Node(nodeID, parentID, leftDaughterID, rightDaughterID, isLeaf, threshold, variableID, outcome);
+    idToNodeMap.insert(IDToNodeMap::value_type(nodeID, pNode));
+
+    return pandora::STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+AdaBoostDecisionTree::Node::Node(const int &nodeID, const int &parentNodeID, const int &leftChildNodeID, const int &rightChildNodeID, const bool &isLeaf, const double &threshold, const int &variableID, const bool outcome) : 
```

Comment:
Check when you need const by reference (big containers, typically) and const by value (built ins, typically), or non-const by reference (calling function wants to receive changed variable).

Reply:
Done.